ember-shortcuts.js

```diff
@@ -60,56 +60,77 @@
     }
   }
 
-  function makeDispatch(router, filters) {
-    function triggerShortcut(def, event) {
-      var i, action, handler, infos;
+  function filter(filters, event) {
+    for (var i = 0; i < filters.length; i++) {
+      if (!filters[i](event)) return false;
+    }
+    return true;
+  }
+
+  function triggerEvent(filters, event, kc, callback) {
+    if (!ENABLED) return;
+    if (!filter(filters, event)) return;
+    if (!(kc in SHORTCUTS)) return;
+
+    forEach(SHORTCUTS[kc], function(def) {
+      if (!isMod(kc) && !modsMatch(def)) return;
+      Ember.run(function() { callback(def, event); });
+    });
+  }
+
+  function makeTriggerShortcut(router, callback) {
+    return function triggerShortcut(def, event) {
+      var actionOrObject, infos;
 
       if (!(infos = router.currentHandlerInfos)) return;
 
-      for (i = infos.length - 1; i >= 0; i--) {
-        handler = infos[i].handler;
+      for (var i = infos.length - 1; i >= 0; i--) {
+        var handler = infos[i].handler;
 
-        if (handler.shortcuts && (action = handler.shortcuts[def.raw])) {
-          handler.send(action, event);
-          return;
+        if (handler.shortcuts && (actionOrObject = handler.shortcuts[def.raw])) {
+          return callback(handler, actionOrObject);
         }
       }
-    }
+    };
+  }
 
-    function filter(event) {
-      for (var i = 0; i < filters.length; i++) {
-        if (!filters[i](event)) return false;
+  function makeKeyDownDispatch(router, filters) {
+    var triggerKeyDownShortcut = makeTriggerShortcut(router, function(handler, actionOrObject) {
+      if (typeof actionOrObject === 'string') {
+        handler.send(actionOrObject, event);
+      } else {
+        handler.send(actionOrObject.keyDown, event);
       }
-      return true;
-    }
+    });
 
-    return function dispatchShortcut(event) {
+    return function dispatchKeyDownShortcut(event) {
       var kc = normalize(event.keyCode);
 
       PRESSED[kc] = true;
-
       if (isMod(kc)) {
         PRESSED_MODS[kc] = true;
-        return;
```

Comment:
Why doesn’t this need the `return` if the pressed key is a modifier logic anymore? Shouldn’t that case work the same as before?

Reply:
See: c60cbd7f6d66aba0f381085d749e67a30dde24c5